llx/builtin_map.go

```diff
@@ -396,6 +395,60 @@ func dictValuesV2(e *blockExecutor, bind *RawData, chunk *Chunk, ref uint64) (*R
 	return ArrayData(res, types.Dict), 0, nil
 }
 
+// Where blocks on strings try to see if the content is found inside the string.
+// Due to the way the compiler behaves, if we check for static values, it is
+// converted into a function:
+//
+//	"hello".contains("ll")  ==>  "hello".contains( _ == "ll" )
+//
+// ie: "hello".where("ll")     ==>  "hello".where(_ == "ll") = "ll"
+//
+// This means that we do not treat these operations the same we would for equality.
+// This is why the operations need a bit of special handling, which is done
+// inside this function.
+func _stringWhere(e *blockExecutor, src string, chunk *Chunk, ref uint64, inverted bool) (*RawData, uint64, error) {
+	arg1 := chunk.Function.Args[1]
+	fref, ok := arg1.RefV2()
+	if !ok {
+		return nil, 0, errors.New("Failed to retrieve function reference of 'where' call")
+	}
+
+	dref, err := e.ensureArgsResolved(chunk.Function.Args[2:], ref)
+	if dref != 0 || err != nil {
+		return nil, dref, err
+	}
+
+	funBlock := e.ctx.code.Block(fref)
+	if len(funBlock.Chunks) == 0 {
+		return BoolFalse, 0, nil
+	}
+
+	// We have to rewrite operations in the block to make them behave correctly
+	// with checking the body of the string, instead of the entire thing.
+	for i := range funBlock.Chunks {
```

Comment:
is this transformation not possible to do in the compiler? Doing it during execution might not be safe if this code is called concurrently. It may also be unexpected to callers to have the struct change from underneath them

Reply:
or maybe a copy of the struct can be made and changes can be made to that

Reply:
Good point, I think I have to move it. 

Unfortunately it is not safe at compile-time, because the behavior changes based on the contents of the dict - which we don't know until it is executed.